Fix range query on date fields for number inputs

cbuescher opened the pull request:
Currently, if you write a date range query with numeric 'to' or 'from' bounds,
they can be interpreted as years if no format is provided. We use
"strict_date_optional_time||epoch_millis" in this case that can interpret inputs
like 1000 as the year 1000 for example. We should change this to always
interpret and parse numbers in this case with the second option "epoch_millis"
if no other formatter was provided.

Closes #63680

Commit:
Fix range quers on date fields for number inputs

Currently, if you write a date range query with numeric 'to' or 'from' bounds,
they can be interpreted as years if no format is provided. We use
"strict_date_optional_time||epoch_millis" in this case that can interpret inputs
like 1000 as the year 1000 for example. We should change this to always
interpret and parse numbers in this case with the second option "epoch_millis"
if no other formatter was provided.

Closes #63680

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

cbuescher commented:
This is a potential fix, as @nik9000 mentioned on the issue we should discuss if we want this kind of change and, if so, should it be considered a breaking change in 8.0 or a bugfix. I tend to see this as a bug fix since the current behaviour is really trappy and very likely not expected by the user.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:
```diff
@@ -349,9 +350,17 @@ public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower
-                    : forcedDateParser;
+            DateMathParser parser;
+            if (forcedDateParser == null) {
+                if (lowerTerm instanceof Integer || upperTerm instanceof Integer) {
```
I think `Number` is better here. That, or `Long` and `Integer`. Its kind of plain luck that the range of numbers that cause the bug are small and so parsed as an `Integer` instead of a `Long` or whatever. But tests and stuff could still end up with a `Long`.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:
```diff
@@ -349,9 +350,17 @@ public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower
+            if (forcedDateParser == null) {
+                if (lowerTerm instanceof Integer || upperTerm instanceof Integer) {
+                    // force epoch_millis
+                    parser = EPOCH_MILLIS_PARSER;
```
Should we just cast them to a number and call `longValue` on them? It feels like a shame to `toString` them and then parse them if we know they are numbers.

nik9000 commented on server/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java:
```diff
@@ -109,4 +111,18 @@ public void testMustRewrite() throws IOException {
         assertEquals("Rewrite first", e.getMessage());
     }
+
+    public void testFoo() {
```
Leftover I think.

Commit:
Iter

Commit:
Merge branch 'master' into fix-63680

Commit:
iter

Commit:
spotless

cbuescher commented on server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:
```diff
@@ -372,25 +381,37 @@ public static Query dateRangeQuery(
-                    l = parseToLong(lowerTerm, !includeLower, timeZone, parser, nowSupplier, resolution);
+                    if (lowerTerm instanceof Number && forcedDateParser == null) {
+                        // force epoch_millis
+                        l = resolution.convert(Instant.ofEpochMilli(((Number) lowerTerm).longValue()));
```
I think we need "resolution.convert" here to this gets kind of ugly. I personally prefer the earlier variant where we just have a "toString" detour but keep the rest of the code as-is.

nik9000 replied:
Yeah. I think you are right. If this were super performance sensitive code I'd argue more, but building queries typically isn't. I think going back is probably the right thing.

cbuescher commented:
@nik9000 thanks for the review, I added another iteration but personally prefered the earlier variant. let me know what you think. Maybe I missed some simpler option?

Commit:
Go back

Commit:
fixing yaml test

cbuescher commented on rest-api-spec/src/main/resources/rest-api-spec/test/field_caps/30_filter.yml:
```diff
@@ -81,7 +81,7 @@ setup:
         index: test-*
         fields: "*"
-        body: { index_filter: { range: { timestamp: { gte: 2010 }}}}
+        body: { index_filter: { range: { timestamp: { gte: "2010" }}}}
```
Interestingly we weren't very precise about this in our own tests.

Commit:
Merge branch 'master' into fix-63680

Commit:
Fix compile problem

Commit:
reapplying yaml fix

Commit:
spotless again

Commit:
Merge branch 'master' into fix-63680

Commit:
remove repeat

nik9000 commented:
This happens when you submit a doc too!

cbuescher commented:
@nik9000 just going through open PRs atm, do you think this needs changes or additions? Or just another round of discussion maybe?

nik9000 reviewed (approved):
LGTM. Should we add something to the breaking changes list?

Commit:
Add docs and breaking changes entry

cbuescher commented:
@nik9000 thanks for the review, I added a blip to the docs and an entry in the breaking changes list. Can you take another short quick look to see if what I added there makes sense to you?

nik9000 reviewed (approved):
LGTM! I think 8.0.0-alpha1.asciidoc could also use a line for this.

Commit:
Adding to breaking changes section in release notes

Commit:
Merge branch 'master' into fix-63680

Commit:
init dateMathParser in DateScriptFieldType
